### PR TITLE
Get rid of `IBranchLayout` and `IBranchLayoutEntry`

### PR DESCRIPTION
**Pull request by PawelLipski**

### CODE REVIEW

**Review by LukaszKontowski — APPROVED** [review]
Makes sense 👍

File: branchLayout/api/src/main/java/com/virtuslab/branchlayout/api/BranchLayoutEntry.java [hunk 1]
```diff
@@ -1,42 +1,67 @@
...
 import lombok.AllArgsConstructor;
 import lombok.Getter;
 import lombok.ToString;
 import lombok.With;
+import lombok.val;
 import org.checkerframework.checker.nullness.qual.Nullable;
 
+/**
+ *  Class that encapsulates a single branch and a list of its children. <br>
+ *  Two IBranchLayoutEntry objects are equal when their names, their custom annotations
```
Comment (mkondratek) [inline branchLayout/api/src/main/java/com/virtuslab/branchlayout/api/BranchLayoutEntry.java:16]: IBranchLayoutEntry

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/base/BaseSlideInBelowAction.java [hunk 2]
```diff
@@ -126,9 +126,9 @@ public void actionPerformed(AnActionEvent anActionEvent) {
     }
 
     // TODO (#430): expose getParent from branch layout API
```
Comment (mkondratek) [inline frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/base/BaseSlideInBelowAction.java:128]: dont we want to close it btw? can we still call it "branch layout API" anyway?

**Review by mkondratek — CHANGES_REQUESTED** [review]

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/base/BaseSlideInBelowAction.java (hunk 2 above)
Comment (PawelLipski) [reply frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/base/BaseSlideInBelowAction.java:128]: #430 is done on #1264, to facilitate #1243

Comment (PawelLipski) [reply frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/base/BaseSlideInBelowAction.java:128]: Let's stick to `branch layout API` naming 🤔

File: backend/impl/src/main/java/com/virtuslab/gitmachete/backend/impl/GitMacheteRepository.java [hunk 3]
```diff
@@ -812,45 +810,46 @@ private static class DiscoverGitMacheteRepositoryAux extends CreateGitMacheteRep
...
-      void removeChild(MyBranchLayoutEntry child) {
+      void removeChild(CompressablePathTreeNode child) {
         children = children.remove(child);
       }
 
-      MyBranchLayoutEntry getRoot() {
-        // Path compression to reduce the lookup time,
-        // see https://en.wikipedia.org/wiki/Disjoint-set_data_structure#Find
+      CompressablePathTreeNode getRoot() {
+        // The actual path compression happens here.
```
Comment (mkondratek) [inline backend/impl/src/main/java/com/virtuslab/gitmachete/backend/impl/GitMacheteRepository.java:852]: nice

**Review by mkondratek — APPROVED** [review]

File: backend/impl/src/main/java/com/virtuslab/gitmachete/backend/impl/GitMacheteRepository.java [hunk 4]
```diff
@@ -872,19 +871,8 @@ private String getRootName() {
...
-      @Override
-      public final boolean equals(@Nullable Object other) {
-        return IBranchLayoutEntry.defaultEquals(this, other);
-      }
-
-      @Override
-      public final int hashCode() {
-        return IBranchLayoutEntry.defaultHashCode(this);
+      public BranchLayoutEntry toBranchLayoutEntry() {
+        return new BranchLayoutEntry(name, /* customAnnotation */ null, children.map(c -> c.toBranchLayoutEntry()));
```
Comment (mkondratek) [inline backend/impl/src/main/java/com/virtuslab/gitmachete/backend/impl/GitMacheteRepository.java:875]: We have a rule:
```
* Don't write nullary lambdas in `receiver::method` notation, use explicit `() -> receiver.method()` notation instead. <br/>
  `::` notation is confusing when applied to parameterless lambdas, as it suggests a unary lambda.
```
and I know that's a one-way implication. 
But why don't we use `::` lambda version here?

Comment (PawelLipski) [reply backend/impl/src/main/java/com/virtuslab/gitmachete/backend/impl/GitMacheteRepository.java:875]: Idk, I just like it better :trollface:  (although both version are okay for unary lambdas)

**Pull request merged**